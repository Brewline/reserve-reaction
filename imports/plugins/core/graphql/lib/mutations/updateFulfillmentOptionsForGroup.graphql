updateFulfillmentOptionsForGroup($input: UpdateFulfillmentOptionsForGroupInput!) {
  updateFulfillmentOptionsForGroup(input: $input) {
    cart {
      _id
    }
  }
}
