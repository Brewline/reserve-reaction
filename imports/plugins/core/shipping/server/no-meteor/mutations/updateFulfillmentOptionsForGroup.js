import { isEqual } from "lodash";
import SimpleSchema from "simpl-schema";
import ReactionError from "@reactioncommerce/reaction-error";
import xformCartGroupToCommonOrder from "/imports/plugins/core/cart/server/no-meteor/util/xformCartGroupToCommonOrder";

import getCartById from "../util/getCartById";

const inputSchema = new SimpleSchema({
  cartId: String,
  cartToken: {
    type: String,
    optional: true
  },
  fulfillmentGroupId: String
});

/**
 * @name getShipmentQuotesQueryStatus
 * @param  {Array} rates    Rate array
 * @return {Object} An object with `shipmentQuotes` and `shipmentQuotesQueryStatus` on it
 * @private
 */
function getShipmentQuotesQueryStatus(rates) {
  if (rates.length === 0) {
    return {
      shipmentQuotes: [],
      shipmentQuotesQueryStatus: {
        requestStatus: "pending"
      }
    };
  }

  const firstRateItem = rates[0];
  if (firstRateItem.requestStatus === "error") {
    return {
      shipmentQuotes: [],
      shipmentQuotesQueryStatus: {
        requestStatus: firstRateItem.requestStatus,
        shippingProvider: firstRateItem.shippingProvider,
        message: firstRateItem.message
      }
    };
  }

  return {
    shipmentQuotes: rates,
    shipmentQuotesQueryStatus: {
      requestStatus: "success",
      numOfShippingMethodsFound: rates.length
    }
  };
}

/**
 * @method updateFulfillmentOptionsForGroup
 * @summary Updates the fulfillment quotes for a fulfillment group
 * @param {Object} context -  an object containing the per-request state
 * @param {Object} input - an object of all mutation arguments that were sent by the client
 * @param {String} input.cartId - The ID of the cart to update fulfillment options for
 * @param {String} [input.cartToken] - The token for the cart, required if it is an anonymous cart
 * @param {String} input.fulfillmentGroupId - The group to update fulfillment options for
 * @return {Promise<Object>} An object with a `cart` property containing the updated cart
 */
export default async function updateFulfillmentOptionsForGroup(context, input) {
  const cleanedInput = inputSchema.clean(input || {});
  inputSchema.validate(cleanedInput);

  const { cartId, cartToken, fulfillmentGroupId } = cleanedInput;
  const { appEvents, collections } = context;
  const { Cart } = collections;

  const cart = await getCartById(context, cartId, { cartToken, throwIfNotFound: true });

  const fulfillmentGroup = (cart.shipping || []).find((group) => group._id === fulfillmentGroupId);
  if (!fulfillmentGroup) throw new ReactionError("not-found", `Fulfillment group with ID ${fulfillmentGroupId} not found in cart with ID ${cartId}`);

  // Map the items onto the fulfillment groups
  fulfillmentGroup.items = fulfillmentGroup.itemIds.map((itemId) => cart.items.find((item) => item._id === itemId));

  const commonOrder = await xformCartGroupToCommonOrder(cart, fulfillmentGroup, context);

  // In the future we want to do this async and subscribe to the results
  const rates = await context.queries.getFulfillmentMethodsWithQuotes(commonOrder, context);

  const { shipmentQuotes, shipmentQuotesQueryStatus } = getShipmentQuotesQueryStatus(rates);

  if (!isEqual(shipmentQuotes, fulfillmentGroup.shipmentQuotes) || !isEqual(shipmentQuotesQueryStatus, fulfillmentGroup.shipmentQuotesQueryStatus)) {
    const { matchedCount } = await Cart.updateOne({
      "_id": cartId,
      "shipping._id": fulfillmentGroupId
    }, {
      $set: {
        "shipping.$.shipmentQuotes": shipmentQuotes,
        "shipping.$.shipmentQuotesQueryStatus": shipmentQuotesQueryStatus
      }
    });
    if (matchedCount !== 1) throw new ReactionError("server-error", "Unable to update cart");

    const updatedCart = await Cart.findOne({ _id: cartId });
    await appEvents.emit("afterCartUpdate", updatedCart);

    return { cart: updatedCart };
  }

  return { cart };
}
