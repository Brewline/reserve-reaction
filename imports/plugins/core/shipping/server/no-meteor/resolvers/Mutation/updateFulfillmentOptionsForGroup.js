import { decodeCartOpaqueId, decodeFulfillmentGroupOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/cart";
import updateFulfillmentOptionsForGroupMutation from "../../mutations/updateFulfillmentOptionsForGroup";

/**
 * @name "Mutation.updateFulfillmentOptionsForGroup"
 * @method
 * @memberof Cart/GraphQL
 * @summary resolver for the updateFulfillmentOptionsForGroup GraphQL mutation
 * @param {Object} parentResult - unused
 * @param {Object} args.input - an object of all mutation arguments that were sent by the client
 * @param {String} args.input.cartId - The ID of the cart to update fulfillment options for
 * @param {String} [args.input.cartToken] - The token for the cart, required if it is an anonymous cart
 * @param {String} args.input.fulfillmentGroupId - The group to update fulfillment options for
 * @param {String} [args.input.clientMutationId] - An optional string identifying the mutation call
 * @param {Object} context - an object containing the per-request state
 * @return {Promise<Object>} UpdateFulfillmentOptionsForGroupPayload
 */
export default async function updateFulfillmentOptionsForGroup(parentResult, { input }, context) {
  const { cartId: opaqueCartId, cartToken, clientMutationId = null, fulfillmentGroupId: opaqueFulfillmentGroupId } = input;

  const fulfillmentGroupId = decodeFulfillmentGroupOpaqueId(opaqueFulfillmentGroupId);
  const cartId = decodeCartOpaqueId(opaqueCartId);

  const { cart } = await updateFulfillmentOptionsForGroupMutation(context, {
    cartId,
    cartToken,
    fulfillmentGroupId
  });

  return {
    cart,
    clientMutationId
  };
}
