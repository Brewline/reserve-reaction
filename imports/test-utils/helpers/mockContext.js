const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
