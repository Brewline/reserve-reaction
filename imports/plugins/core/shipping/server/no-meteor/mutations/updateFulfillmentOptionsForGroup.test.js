import Factory from "/imports/test-utils/helpers/factory";
import mockContext from "/imports/test-utils/helpers/mockContext";
import updateFulfillmentOptionsForGroup from "./updateFulfillmentOptionsForGroup";

jest.mock("../util/getCartById", () => jest.fn().mockImplementation(() => Promise.resolve({
  _id: "cartId",
  items: [{
    _id: "123"
  }],
  shipping: [{
    _id: "group1",
    itemIds: ["123"],
    type: "shipping"
  }]
})));

const fakeCart = Factory.Cart.makeOne();
const fakeQuote = Factory.ShipmentQuote.makeOne();
const mockGetFulfillmentMethodsWithQuotes = jest.fn().mockName("getFulfillmentMethodsWithQuotes");

beforeAll(() => {
  mockContext.queries = {
    getFulfillmentMethodsWithQuotes: mockGetFulfillmentMethodsWithQuotes
  };
});

beforeEach(() => {
  mockGetFulfillmentMethodsWithQuotes.mockClear();
});

test("updates cart properly for empty rates", async () => {
  mockGetFulfillmentMethodsWithQuotes.mockReturnValueOnce(Promise.resolve([]));
  mockContext.collections.Cart.findOne.mockReturnValueOnce(Promise.resolve(fakeCart));

  const result = await updateFulfillmentOptionsForGroup(mockContext, {
    cartId: "cartId",
    fulfillmentGroupId: "group1"
  });
  expect(result).toEqual({ cart: fakeCart });

  expect(mockContext.collections.Cart.updateOne).toHaveBeenCalledWith({
    "_id": "cartId",
    "shipping._id": "group1"
  }, {
    $set: {
      "shipping.$.shipmentQuotes": [],
      "shipping.$.shipmentQuotesQueryStatus": { requestStatus: "pending" }
    }
  });
});

test("updates cart properly for error rates", async () => {
  mockGetFulfillmentMethodsWithQuotes.mockReturnValueOnce(Promise.resolve([{
    requestStatus: "error",
    shippingProvider: "all",
    message: "All requests for shipping methods failed."
  }]));
  mockContext.collections.Cart.findOne.mockReturnValueOnce(Promise.resolve(fakeCart));

  const result = await updateFulfillmentOptionsForGroup(mockContext, {
    cartId: "cartId",
    fulfillmentGroupId: "group1"
  });
  expect(result).toEqual({ cart: fakeCart });

  expect(mockContext.collections.Cart.updateOne).toHaveBeenCalledWith({
    "_id": "cartId",
    "shipping._id": "group1"
  }, {
    $set: {
      "shipping.$.shipmentQuotes": [],
      "shipping.$.shipmentQuotesQueryStatus": {
        requestStatus: "error",
        shippingProvider: "all",
        message: "All requests for shipping methods failed."
      }
    }
  });
});

test("updates cart properly for success rates", async () => {
  mockGetFulfillmentMethodsWithQuotes.mockReturnValueOnce(Promise.resolve([fakeQuote]));
  mockContext.collections.Cart.findOne.mockReturnValueOnce(Promise.resolve(fakeCart));

  const result = await updateFulfillmentOptionsForGroup(mockContext, {
    cartId: "cartId",
    fulfillmentGroupId: "group1"
  });
  expect(result).toEqual({ cart: fakeCart });

  expect(mockContext.collections.Cart.updateOne).toHaveBeenCalledWith({
    "_id": "cartId",
    "shipping._id": "group1"
  }, {
    $set: {
      "shipping.$.shipmentQuotes": [fakeQuote],
      "shipping.$.shipmentQuotesQueryStatus": {
        requestStatus: "success",
        numOfShippingMethodsFound: 1
      }
    }
  });
});

test("throws if there is no fulfillment group with the given ID", async () => {
  await expect(updateFulfillmentOptionsForGroup(mockContext, {
    cartId: "cartId",
    fulfillmentGroupId: "group2"
  })).rejects.toThrowErrorMatchingSnapshot();
});
