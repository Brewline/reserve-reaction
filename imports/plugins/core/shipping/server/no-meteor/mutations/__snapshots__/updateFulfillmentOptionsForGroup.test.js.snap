// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if there is no fulfillment group with the given ID 1`] = `"Fulfillment group with ID group2 not found in cart with ID cartId"`;
